bot/handlers/role_selection.go

```diff
@@ -92,97 +86,115 @@ func (r *RoleSelectionHandler) ReconcileConfig(s *discordgo.Session) error {
 					continue
 				}
 			}
+		} else {
+			r.prompts[existingMessage.ID] = p
+			for _, op := range p.Options {
+				var reactionExists bool
+				for _, reaction := range existingMessage.Reactions {
+					if reaction.Emoji.ID == op.EmojiID {
+						reactionExists = true
+						break
+					}
+				}
+				if !reactionExists {
+					err := s.MessageReactionAdd(existingMessage.ChannelID, existingMessage.ID, op.EmojiID)
+					if err != nil {
+						r.logger.Error("error adding reaction",
+							zap.String("channelName", roleChannel.Name),
+							zap.String("channelId", roleChannel.ID),
+							zap.String("emojiID", op.EmojiID),
+							zap.String("option", op.Description),
+							zap.Error(err))
+						continue
+					}
+				}
+			}
 		}
 	}
+	r.initialized = true
 	return nil
 }
 
-// HandleReactionAdd handles a reaction add event.
-func (r *RoleSelectionHandler) HandleReactionAdd(s *discordgo.Session, m *discordgo.MessageReactionAdd) {
+// handleReaction is a helper function to handle both reaction add and remove events.
+func (r *RoleSelectionHandler) handleReaction(s *discordgo.Session, guildID, channelID, messageID, userID, emojiID string) (string, *discordgo.Member, error) {
+	r.reconciling.RLock()
 	if !r.initialized {
-		err2 := r.ReconcileConfig(s)
-		if err2 != nil {
-			r.logger.Error("failed to reconcile config", zap.Error(err2))
-		} else {
-			r.initialized = true
+		r.reconciling.RUnlock()
+		if err := r.ReconcileConfig(s); err != nil {
+			r.logger.Error("failed to reconcile config", zap.Error(err))
+			return "", nil, err
 		}
+		r.reconciling.RLock()
 	}
-	// Ignore reactions in non-target channel
-	if m.ChannelID != r.config.Channel.ID {
-		return
+
+	if channelID != r.config.Channel.ID {
+		r.reconciling.RUnlock()
+		return "", nil, nil // Not the channel we're watching
 	}
-	if pr, ok := r.prompts[m.MessageID]; !ok {
-		r.logger.Debug("message is not registered prompt", zap.String("message", m.MessageID))
-	} else {
-		var focusRoleId string
-		for _, op := range pr.Options {
-			if op.EmojiID == m.Emoji.ID {
-				focusRoleId = op.RoleID
-			}
-		}
-		if focusRoleId == "" {
-			// unrelated emoji?
-			return
-		}
-		user, err := s.GuildMember(m.GuildID, m.UserID)
-		if err != nil {
-			r.logger.Error("error fetching guild member", zap.Error(err))
-		}
-		for _, roleID := range user.Roles {
-			if roleID == focusRoleId {
-				// user has role
-				return
-			}
+
+	pr, ok := r.prompts[messageID]
+	r.reconciling.RUnlock()
+	if !ok {
```

Comment:
@jules you should add a debug logger here for future reference similar to the previous one